src/looprestoration.rs

```diff
@@ -23,3 +24,115 @@ pub union LooprestorationParams {
     pub filter: Align16<[[int16_t; 8]; 2]>,
     pub sgr: LooprestorationParams_sgr,
 }
+
+type pixel = libc::c_void;
+pub type const_left_pixel_row = *const libc::c_void; // *const [pixel; 4]
+
+pub type looprestorationfilter_fn = Option<
+    unsafe extern "C" fn(
+        *mut pixel,
+        ptrdiff_t,
+        const_left_pixel_row,
+        *const pixel,
+        libc::c_int,
+        libc::c_int,
+        *const LooprestorationParams,
+        LrEdgeFlags,
+        libc::c_int,
+    ) -> (),
+>;
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Dav1dLoopRestorationDSPContext {
+    pub wiener: [looprestorationfilter_fn; 2],
+    pub sgr: [looprestorationfilter_fn; 3],
+}
+
+macro_rules! extern_fn {
+    ( $( $name:ident, )* ) => {
+        extern "C" {
+            $(
+                // TODO(randomPoison): Temporarily pub until init fns are deduplicated.
+                pub(crate) fn $name(
+                    dst: *mut pixel,
+                    dst_stride: ptrdiff_t,
+                    left: const_left_pixel_row,
+                    lpf: *const pixel,
+                    w: libc::c_int,
+                    h: libc::c_int,
+                    params: *const LooprestorationParams,
+                    edges: LrEdgeFlags,
+                    bitdepth_max: libc::c_int,
+                );
+            )*
+        }
+    };
+}
+
+#[cfg(all(
+    feature = "bitdepth_8",
+    feature = "asm",
+    any(target_arch = "x86", target_arch = "x86_64"),
+))]
+extern_fn! {
+    dav1d_wiener_filter7_8bpc_sse2,
+    dav1d_wiener_filter5_8bpc_sse2,
```

Comment:
To help with @thedataking's concerns for using a macro for this (https://github.com/memorysafety/rav1d/pull/332#discussion_r1267657889), how about we make the macro something like `decl_looprestoration_fn!(fn dav1d_wiener_filter7_8bpc_sse2);`?  This way it's
* closer to the C version
* declares one `fn` per macro call, so go-to-definition is easier to use
* keeps the `fn $name` together so that it's more easily greppable

Reply:
I don't think there's a big advantage to only declaring one fn per macro invocation, and I think it mostly increases visual noise. However, I do think doing `fn $name` is a good idea in order to make it easier to find the fn declarations by grepping 👍

Reply:
Appreciate the effort to address my concerns. That said, what does the macro buy you? Code compactness and not much else. Code compactness takes a backseat to readability, debuggability, greppability, etc. IMHO.

I often grep for `fn dav1d_wiener_filter7_8bpc_sse2` or some variation thereof, maybe `fn dav1d_wiener_filter7_8bpc` if I only want to see function definitions. Please (please), consider using macros sparingly. C uses macros and typedefs extensively as a way to paper over portability issues. Rust often provides portable types and APIs so the reasonable expectation is for Rust code to use macros less.

A good use of macros is when you have a "can't see the forrest for all the trees" situation where you can't see the overall pattern of the code for the sheer repetitiveness. I think we had situations like that in `decode.rs` where using macros helped. I just don't think we have the same situation here.

Thanks for hearing out my concerns. Now that I've voiced my concerns, I'll extricate myself from this discussion.

Reply:
Yeah, it seems like rust-analyzer is able to figure it out in this case, likely because the macro is so simple.

Another reason to keep the current version is that it makes it easier to group fns by their `#[cfg]` attribute. Doing one macro invocation per function means that the `#[cfg]` attributes would need to be duplicated a lot, which hurts readability.

I think what I have currently is cleaner than the proposed alternative, but if you feel strongly that we should go with a different approach then I can make the change.

Reply:
> Another reason to keep the current version is that it makes it easier to group fns by their `#[cfg]` attribute. Doing one macro invocation per function means that the `#[cfg]` attributes would need to be duplicated a lot, which hurts readability.

I hadn't thought of that.  That definitely helps a lot then.  But the macro invocation per function could also be done by putting the `extern "C"` block outside of the macro invocations, where we could add the `#[cfg]`s.

Reply:
>  if you feel strongly that we should go with a different approach then I can make the change

I'm fine with what you have, keeping my LGTM. I wasn't factoring in that the macro helped you group fns by cfg. Also good to know that grepping and navigation is expected to work 👍

Reply:
> Appreciate the effort to address my concerns. That said, what does the macro buy you? Code compactness and not much else. Code compactness takes a backseat to readability, debuggability, greppability, etc. IMHO.
> 
> I often grep for `fn dav1d_wiener_filter7_8bpc_sse2` or some variation thereof, maybe `fn dav1d_wiener_filter7_8bpc` if I only want to see function definitions. Please (please), consider using macros sparingly. C uses macros and typedefs extensively as a way to paper over portability issues. Rust often provides portable types and APIs so the reasonable expectation is for Rust code to use macros less.
> 
> A good use of macros is when you have a "can't see the forrest for all the trees" situation where you can't see the overall pattern of the code for the sheer repetitiveness. I think we had situations like that in `decode.rs` where using macros helped. I just don't think we have the same situation here.

To me, the macro here is more readable because it lets you clearly see that every function has the same signature.  That's obscured when you see a wall of code.  There's not that much use in reading each function individually as they're all just variations of the same function.  The macros we had in the devirtualization PRs were a lot more complex, but I think this one is very simple and makes it easier for me to read, and I think for @randomPoison, too.